source/cbor.c

```diff
@@ -0,0 +1,739 @@
+                result.required);
+            decoder->error_code = AWS_ERROR_INVALID_CBOR;
+            break;
+        case CBOR_DECODER_ERROR:
+            AWS_LOGF_ERROR(AWS_LS_COMMON_CBOR, "The cbor data is malformed to decode.");
+            decoder->error_code = AWS_ERROR_INVALID_CBOR;
+            break;
+        default:
+            break;
+    }
+
+    if (decoder->error_code) {
+        /* Error happened during decoding */
+        return aws_raise_error(decoder->error_code);
+    }
+
+    aws_byte_cursor_advance(&decoder->src, result.read);
+    decoder->bytes_decoded += result.read;
+
+    return AWS_OP_SUCCESS;
+}
+
+#define GET_NEXT_ITEM(field, out_type, expected_cbor_type)                                                             \
+    /* NOLINTNEXTLINE(bugprone-macro-parentheses) */                                                                   \
+    int aws_cbor_decoder_pop_next_##field(struct aws_cbor_decoder *decoder, out_type *out) {                           \
+        if ((decoder)->error_code) {                                                                                   \
+            /* Error happened during decoding */                                                                       \
+            return aws_raise_error((decoder)->error_code);                                                             \
+        }                                                                                                              \
+        if ((decoder)->cached_context.type != AWS_CBOR_TYPE_MAX) {                                                     \
+            /* There was a cached context, check if the cached one meets the expected. */                              \
+            goto decode_done;                                                                                          \
+        }                                                                                                              \
+        if (s_cbor_decode_next_element(decoder)) {                                                                     \
+            return AWS_OP_ERR;                                                                                         \
+        }                                                                                                              \
+    decode_done:                                                                                                       \
+        if ((decoder)->cached_context.type != (expected_cbor_type)) {                                                  \
+            AWS_LOGF_ERROR(                                                                                            \
+                AWS_LS_COMMON_CBOR,                                                                                    \
+                "The decoder got unexpected type: %d, while expecting type: %d.",                                      \
```

Comment:
might be worth a `const char *aws_cbor_type_cstr(aws_cbor_type type)` helper function, that you can use here, so the error message is more understandable to otehrs. Similar to https://github.com/awslabs/aws-c-io/blob/375f393ba6b900e7c3a46f9f5a0eebf2f7a22073/source/pkcs11_lib.c#L301-L309

"The decoder got unexpected type: 0, while expecting type: 2."
vs
"The decoder got unexpected type: 0 (AWS_CBOR_TYPE_UINT), while expecting type: 2 (AWS_CBOR_TYPE_FLOAT)."